docs/cluster_api.md

```diff
@@ -0,0 +1,121 @@
+    - `PodNetworkCIDR [array of string]` *Subnet for pod network (osm_cluster_network_cidr in ansible).*
+- `MachineSets [array of ClusterMachineSet]` *List of MachineSets to create for this cluster*
+  - `Name [string, required]` *Name for the machine set - only used to identify the machine set within the cluster. Must be unique within the cluster*
+  - `MachineSetConfig [MachineSetConfig,inline]` *Configuration for the machine set -- see below*
+
+#### Status:
+- `AdminKubeconfig [corev1.LocalReference]` *Reference to secret that contains an admin.kubeconfig to access the cluster as cluster administrator after the cluster is up and running*
+- `Conditions [array]` *Detailed status for the cluster*
+  - `Type [string]` *Type of condition - InfraProvisioned, MasterConfigured, Running*
+  - `Status [string]` *Status of the condition - True, False, Unknown*
+  - `LastProbeTime [string]` *-optional- last time the condition was checked*
+  - `LastTransitionTime [string]` *-optional- last time the condition transitioned*
+  - `Message [string]` *-optional- Human readable message associated with the condition*
+  - `Reason [string]` *-optional- Single word camel-case reason for the condition*
+- `MachineSetCount [int]` *count of machine sets that have been created for this cluster*
+- `Provisioned [boolean]` *true if cluster-level infrastructure has been provisioned for this cluster*
+- `Running [boolean]` *true if the cluster is running and is ready to accept API requests* 
+- `MasterMachineSetName [string]` *name of the machine set designated as master*
+- `InfraMachineSetName [string]` *name of the machine set designated as infra (could be the same as master)*
+
+### MachineSet
+
+- `ObjectMetadata [metav1.ObjectMetadata]`
+- `TypeMetadata [metav1.TypeMetadata]`
+
+#### Spec
+- `MachineSetConfig [MachineSetConfig,inline]` *Configuration for the machine set (see below)*
+
+#### Status
+- `NodesProvisioned [int]` *Number of nodes provisioned by the cloud provider*
+- `NodesReady [int]` *Number of nodes ready*
+- `Conditions [array]` *Detailed status*
+  - `Type [string]` *Type of condition - Provisioned, Ready, etc*
+  - `Status [string]` *Status of the condition - True, False, Unknown*
+  - `LastProbeTime [string]` *-optional- last time the condition was checked*
+  - `LastTransitionTime [string]` *-optional- last time the condition transitioned*
+  - `Message [string]` *-optional- Human readable message associated with the condition*
+  - `Reason [string]` *-optional- Single word camel-case reason for the condition*
+
+
+### MachineSetConfig
+  - `NodeType [NodeType:string]` *Enumerated node type for the machine set (compute or master)*
+  - `Infra [boolean]` *True if this machine set should contain infrastructure pods (registry, router)*
```

Comment:
The controller needs to know the node type to determine whether to run the master playbook on the node. Consequently, at least for the master nodes, the value needs to be there, whether it is in a label or in it's own field. It doesn't seem appropriate to me to require the user to know that certain labels are required. What is the benefit of having the information in a label instead of a field? What is the reason for resistance to a field?

Reply:
My main opposition to fields stems largely from the fact that infra nodes are an openshift only concept that don't actually make any sense to me in the greater picture of what we are trying to build. There should never be any special magic nodes. All pinning that we do to special nodes is wrong. I understand why we do some of that today (mostly around the lack of a deschedular, the fact that we don't really trust resource usage separation, and difficulties getting traffic to our routers), but codifying specialness into our API for all time seems very wrong. I want infra nodes to die.

Masters are a little bit different. Masters themselves wouldn't exist at all in a self-hosted world (aka, what I think bootkube ends up doing) but I think the path to success for having non-special masters is long enough are far enough out of my vision that I'm ok codifying masters into the api.

All of this is also colored with the fact that I believe we MUST merge with the cluster-api working group. They are starting from the bottom and working up. We are starting from the top and working down. But eventually we must go to them. Upstream has NO concept of an infra. Infra will not survive the merger.

Reply:
It sounds like we are conflating issues a bit.

`NodeType` as it is now only distinguishes between machines that have the master setup run on them and those that do not. It has nothing to do with infra in any way. Writing that out makes it striking to me that a boolean would be a better choice than an enumeration. If we have a boolean that identified whether a node was a master node or not, would you have an objection to that? In terms of staying in line with upstream, the current iteration of that cluster API has a string array of roles for each machine. Valid roles currently are "Master" and "Node". So, I don't think that we are far apart from upstream with what we have here, with respect to node types.

For determining which machines have infra setup run on them, I could see that fitting into the roles concept from upstream, where another role name "Infra" is added to the current roles of "Master" and "Node". Perhaps the roles can be extensible beyond the kubernetes-wide roles of "Master" and "Node". Openshift could accept a role of "Infra" that is ignored or not accepted by other platforms.

Reply:
+1 to the idea of having roles. @eparis would that be acceptable?
It would give me enough flexibility to create a cluster with a single machine set where the machine set has the role of master, node and infra. And the infra role is something upstream doesn't have to know about.